### PR TITLE
Moving build from podman to a BuildConfig

### PR DESCRIPTION
**Pull request by csibbitt**
* Moving to an OpenShift native BuildConfig to build the SFO
  * It turns out operator-sdk is not really needed for this. :P
* It doesn't require podman or any special perms in the build container!
  * Just access to the openshift API to create the build object
* Cleaned up separation of build, deploy, and test

### CODE REVIEW

**Comment by csibbitt** [comment]
[Investigate zuul, dciy, gitlab, ocp jenkins, etc to see if they smoothe the path forward](https://trello.com/c/OtdhAr9S/902-investigate-zuul-dciy-gitlab-ocp-jenkins-etc-to-see-if-they-smoothe-the-path-forward)

**Comment by csibbitt** [comment]
# Initial Testing

This work isn't done, but the build itself appears to work just fine.

## BuildConfig POC
```
$ oc new-project buildtest
[...]

$ oc get is
No resources found.

$ ./build.sh 
Error from server (NotFound): imagestreams.image.openshift.io "service-assurance-operator" not found
imagestream.image.openshift.io/service-assurance-operator created
Error from server (NotFound): buildconfigs.build.openshift.io "service-assurance-operator" not found
buildconfig.build.openshift.io/service-assurance-operator created
build.build.openshift.io/service-assurance-operator-1 started

$ oc get build
NAME                           TYPE     FROM          STATUS     STARTED              DURATION
service-assurance-operator-1   Docker   Git@8dc707a   Complete   About a minute ago   56s

$ oc get is
NAME                         IMAGE REPOSITORY                                                                               TAGS     UPDATED
service-assurance-operator   default-route-openshift-image-registry.apps-crc.testing/buildtest/service-assurance-operator   latest   About a minute ago
```

Neat! So I build the docker image without using operator-sdk, which solves all of the podmad issues I had. :P

## Does it work in CI?

```
[csibbitt@laptop build (csibbitt-902-buildconfig) (⎈ |default/api-crc-testing:6443/kube:admin)]$ ./build_ci.sh 
[...]
* [info] Building...

Error from server (NotFound): imagestreams.image.openshift.io "service-assurance-operator" not found
imagestream.image.openshift.io/service-assurance-operator created
Error from server (NotFound): buildconfigs.build.openshift.io "service-assurance-operator" not found
buildconfig.build.openshift.io/service-assurance-operator created
build.build.openshift.io/service-assurance-operator-1 started

* [info] Re-deploying with local build...

subscription.operators.coreos.com "serviceassurance-operator-alpha-redhat-service-assurance-operators-openshift-marketplace" deleted
clusterserviceversion.operators.coreos.com "service-assurance-operator.v0.1.1" deleted
serviceaccount/service-assurance-operator created
role.rbac.authorization.k8s.io/service-assurance-operator created
rolebinding.rbac.authorization.k8s.io/service-assurance-operator created
clusterserviceversion.operators.coreos.com/service-assurance-operator.v0.1.1 created
serviceassurance.infra.watch/saf-default created

* [info] CI Build complete. You can now run tests.

[csibbitt@laptop build (csibbitt-902-buildconfig) (⎈ |sa-telemetry/api-crc-testing:6443/kube:admin)]$ oc describe po service-assurance-operator-c8b7f986f-8ghtb | grep Image
                    containerImage: quay.io/redhat-service-assurance/service-assurance-operator
    Image:         image-registry.openshift-image-registry.svc:5000/sa-telemetry/service-assurance-operator:latest
    Image ID:      image-registry.openshift-image-registry.svc:5000/sa-telemetry/service-assurance-operator@sha256:4bfa5b30d726f207fae25ad30bcd19958e096c2a8d8e4dd225c38f6e21900a9e
    Image:          image-registry.openshift-image-registry.svc:5000/sa-telemetry/service-assurance-operator:latest
    Image ID:       image-registry.openshift-image-registry.svc:5000/sa-telemetry/service-assurance-operator@sha256:4bfa5b30d726f207fae25ad30bcd19958e096c2a8d8e4dd225c38f6e21900a9e
  Warning  Failed     3m54s (x3 over 4m20s)  kubelet, crc-rk2fc-master-0  Error: ImagePullBackOff
  Warning  Failed     3m54s (x5 over 4m20s)  kubelet, crc-rk2fc-master-0  Error: ImagePullBackOff
  Warning  Failed     3m40s (x3 over 4m20s)  kubelet, crc-rk2fc-master-0  Error: ErrImagePull


[csibbitt@laptop build (csibbitt-902-buildconfig) (⎈ |sa-telemetry/api-crc-testing:6443/kube:admin)]$ ../tests/smoketest/smoketest.sh 
[...]
*** [SUCCESS] Smoke test job completed successfully
```

Yup! It works!

Note the "ErrImagePull" Warnings. This is because we don't wait between triggering the build and redeploying the patched CSV that uses it. It will retry until the image appears in the registry, or eventually our test job will time out. In addition to this, I moved the build step AFTER the quickstart.sh call because you need a namespace to exist before you can run a Build in it - this means we don't even attempt the build until after the first trunk deployment. So overall this is cleaner but quite a bit slower to fail than the previous build strategy. If it helps us align more closely with other OpenShift projects, it's worth it, though.

**Comment by pleimer** [comment]
Awesome looks great!

**Comment by csibbitt** [comment]
Final version hasn't been tested end-to-end, but feel free to give it a shot (or maybe CI bot will do it for me?) 

I especially like this part. Nice and clean: 
``
        - ./build/build.sh
        - ./build/deploy_ci.sh
        - ./tests/smoketest/smoketest.sh
```

@pleimer this also eliminates all of the builder service account funny business requirements and podman and stuff from the ci server container. I'll tidy up that side once this merges.

File: ci.yml [hunk 1]
```diff
@@ -1,7 +1,10 @@
 script:
-        - oc login -u kubeadmin -p mhk2X-Y8ozE-9icYb-uLCdV --insecure-skip-tls-verify=true https://api.crc.testing:6443
-        - IMAGE_BUILD_ARGS="--storage-driver vfs" ./build/build_ci.sh
-        - cd ./tests/smoketest 
-        - ./smoketest.sh
+        - export HOME=$(dirname "$PWD")
+        - oc login -u kubeadmin -p BdTY8-Vra64-gD3cI-dAPS6 --insecure-skip-tls-verify=true https://api.crc.testing:6443
+        - oc new-project "${OCP_PROJECT}"
+        - ./build/build.sh
+        - ./build/deploy_ci.sh
```
Comment (leifmadsen) [inline ci.yml]: I think you want `./build/build_ci.sh` here.

Comment (leifmadsen) [reply ci.yml]: Nevermind. I did look initially for a filename change, but missed it. I see it now.

**Comment by csibbitt** [comment]
CI Failed; it looks like it never succeeded in instantiating the components after the local build (operators are there but none of the local components). I'll re-test locally and see if I can spot why.

**Comment by csibbitt** [comment]
Worked locally, but doesn't work in the CI Bot. At least two problems: 
* Build itself fails with `error: the directory "." is not a valid Git repository: fatal: not a git repository (or any of the parent directories): .git`  which is a bit of a surprise
* The fact that the build failed should have caused the entire test to fail, but it continues on

**Comment by csibbitt** [comment]
@pleimer @leifmadsen FYI this is passing CI now (love being able to say that again!), but we want to merge https://github.com/redhat-service-assurance/service-assurance-operator/pull/44 first and fix this on top of it, so I will not merge it yet.

**Comment by csibbitt** [comment]
I fixed a single conflicted line (to take your value) and this seems to be good to go! Let's see what the bot says

**Pull request merged**